akka-actor/src/main/scala/akka/actor/FSM.scala

```diff
@@ -648,13 +653,18 @@ trait FSM[S, D] extends Actor with Listeners with ActorLogging {
         this.nextState = null
       }
       currentState = nextState
-      val timeout = if (currentState.timeout.isDefined) currentState.timeout else stateTimeouts(currentState.stateName)
-      if (timeout.isDefined) {
-        val t = timeout.get
-        if (t.isFinite && t.length >= 0) {
-          import context.dispatcher
-          timeoutFuture = Some(context.system.scheduler.scheduleOnce(t, self, TimeoutMarker(generation)))
-        }
+
+      def scheduleTimeout(d: FiniteDuration): Some[Cancellable] = {
+        import context.dispatcher
+        Some(context.system.scheduler.scheduleOnce(d, self, TimeoutMarker(generation)))
+      }
+
+      currentState.timeout match {
+        case SomeMaxFiniteDuration                    ⇒ // effectively disable stateTimeout
```

Comment:
this is just a workaround that the original code did not handle Duration.Inf although the forMax method (L146) accepts a Duration. The proper fix would be to change the signature of State to have a timeout of Option[Duration] instead of Option[FiniteDuration], but that would break binary compatiblility. So Konrad introduced this artifical marker that is introduced at L148 when a Duration.Inf is encountered, and which can be stored in the State case class without changing its signature.